core/src/docs/rfcs/6209_glob_support.md

```diff
@@ -0,0 +1,167 @@
+- Proposal Name: `glob_support`
+- Start Date: 2025-05-21
+- RFC PR: [apache/opendal#6209](https://github.com/apache/opendal/pull/6209)
+- Tracking Issue: [apache/opendal#6210](https://github.com/apache/opendal/issues/6210)
+
+# Summary
+
+Add support for matching file paths against Unix shell style patterns (glob) in OpenDAL.
+
+# Motivation
+
+Glob patterns are a widely used way to filter files based on their paths. They provide a simple and intuitive syntax for matching multiple files with similar path patterns. Adding glob support to OpenDAL would enable users to easily filter and process files that match certain patterns without having to implement this functionality themselves.
+
+Currently, users who want to filter objects based on patterns have to list all objects and then apply filters manually, which is verbose and not very intuitive. By providing native glob support, we can make this common operation more convenient and efficient.
+
+# Guide-level explanation
+
+With glob support, users can easily match files based on patterns. The API would be available as an option on the `list_with` method, allowing users to get an iterator over entries that match the provided glob pattern.
+
+For example:
+
+```rust
+// Get an iterator over all jpeg files in the media directory and its subdirectories
+let it = op.list_with("media/").glob("**/*.jpg").await?;
+
+while let Some(entry) = it.next().await? {
+   do_something(&entry);
+}
+```
+
+The glob syntax would support common patterns like:
+
+- `*` - Match any sequence of non-separator characters
+- `?` - Match any single non-separator character
+- `**` - Match any sequence of characters including separators
+- `{a,b}` - Match either a or b
+- `[ab]` - Match either a or b
+- `[a-z]` - Match any character in range a-z
+
+Both blocking and non-blocking (async) versions of the API would be provided:
+
+```rust
+// Async version
+async fn list_with(&self, path: &str) -> Lister {
+    // ...
+}
+
+impl Lister {
+    pub fn glob(self, pattern: &str) -> Result<impl Stream<Item = Result<Entry>>> {
```

Comment:
Hi, I feel that adding an API for `Lister` doesn't looks good.

For example, users can use `Lister` like:

```rust
let l = op.lister(path).await?;
let next_entry = l.next().await?;
let glob = l.glob(path).await?;
```

The results of `glob` returns can be confusing for users.

My current idea is that we add `glob` as a new option for `list`:

```rust
let entries: Vec<Entry> = op.list_with(path).glob("**.parquet").await?;
let lister: Lister = op.lister_with(path).glob("**.parquet").await?;
```